src/ducks/transactions/TransactionRow/LabelField.jsx

```diff
@@ -0,0 +1,130 @@
+import {
+  getLabel,
+  hasEditedLabel,
+  setTransactionLabel
+} from 'ducks/transactions/helpers'
+
+import styles from 'ducks/transactions/TransactionRow/LabelField.styl'
+
+export default function LabelField({ transaction, variant = 'body1' }) {
+  const client = useClient()
+  const [isEditing, setIsEditing] = useState(false)
+  const [isSaving, setIsSaving] = useState(false)
+  const inputRef = useRef()
+  const containerRef = useRef()
+
+  const saveAndClose = useCallback(
+    async label => {
+      if (
+        label !== transaction.manualLabel &&
+        label !== getLabel(transaction, true)
+      ) {
+        const sanitized = label?.trim()
+        try {
+          setIsSaving(true)
+          await client.save(
+            setTransactionLabel(
+              transaction,
+              sanitized?.length ? sanitized : null
+            )
+          )
+        } finally {
+          setIsSaving(false)
+        }
+      }
+      setIsEditing(false)
+    },
+    [client, setIsEditing, setIsSaving, transaction]
+  )
+
+  const handleEdit = useCallback(
+    event => {
+      event.stopPropagation()
+      setIsEditing(true)
+    },
+    [setIsEditing]
+  )
+
+  const handleClose = useCallback(
+    () => saveAndClose(inputRef.current.value),
+    [saveAndClose]
+  )
+
+  const handleRestore = useCallback(() => {
+    saveAndClose(null)
+  }, [saveAndClose])
+
+  const handleKeyDown = useCallback(
+    ({ key }) => {
+      if (key === 'Enter') {
+        handleClose()
+      } else if (key === 'Escape') {
+        setIsEditing(false)
+      }
+    },
+    [handleClose, setIsEditing]
+  )
+
+  const handleBlur = useCallback(
+    ({ relatedTarget }) => {
+      if (containerRef.current?.contains(relatedTarget)) {
+        // do not close when focused element is inside the component
+        return
+      }
+      saveAndClose(inputRef.current.value)
+    },
+    [saveAndClose]
+  )
+
+  return isEditing ? (
+    <div
+      onBlur={handleBlur}
+      ref={containerRef}
+      onClick={e => e.stopPropagation()}
+    >
+      <TextField
+        autoFocus
+        fullWidth
+        defaultValue={getLabel(transaction)}
+        className={styles.EditableLabel}
+        inputRef={inputRef}
+        onKeyDown={handleKeyDown}
+        variant="outlined"
+        InputProps={{
+          endAdornment: isSaving ? (
+            <CircularProgress size={24} />
+          ) : (
+            <>
+              <IconButton size="small" onClick={handleClose}>
+                <Icon icon={CrossCircleOutlineIcon} size={14} />
+              </IconButton>
+              {hasEditedLabel(transaction) ? (
+                <IconButton size="small" onClick={handleRestore}>
+                  <Icon icon={RestoreIcon} size={14} />
+                </IconButton>
+              ) : null}
+            </>
+          )
+        }}
+      />
+    </div>
+  ) : (
+    <div className={styles.FixedLabel}>
```

Comment:
nit: We could replace this classes with utility classes like `u-flex u-flex-items-center` and `u-ml-haft` on the `IconButton`